services/user/Tokens/src/RTokens.cpp

```diff
@@ -13,64 +15,229 @@ using namespace UserService;
 using namespace std;
 using namespace psibase;
 
-auto tokenService = QueryableService<Tokens::Tables, Tokens::Events>{Tokens::service};
-
-struct UserBalanceRecord
+auto tokenService = Tokens::Tables{Tokens::service};
+namespace TokenQueryTypes
 {
-   AccountNumber user;
-   uint64_t      balance;
-   uint8_t       precision;
-   TID           token;
-   SID           symbol;
-};
-PSIO_REFLECT(UserBalanceRecord, user, balance, precision, token, symbol);
+   auto realBalance = [](uint64_t bal, const Precision& p) -> uint64_t
+   { return static_cast<uint64_t>(bal / pow(10, p.value)); };
+
+   struct Credit
+   {
+      SID           symbolId;
+      TID           tokenId;
+      Precision     precision;
+      uint64_t      balance;
+      AccountNumber creditedTo;
+
+      auto getKey() const { return realBalance(balance, precision); }
+   };
+   PSIO_REFLECT(Credit, symbolId, tokenId, precision, balance, creditedTo);
+
+   struct Debit
+   {
+      SID           symbolId;
+      TID           tokenId;
+      Precision     precision;
+      uint64_t      balance;
+      AccountNumber debitableFrom;
+
+      auto getKey() const { return realBalance(balance, precision); }
+   };
+   PSIO_REFLECT(Debit, symbolId, tokenId, precision, balance, debitableFrom);
+
+   struct TokenBalance
+   {
+      SID       symbolId;
+      TID       tokenId;
+      Precision precision;
+      uint64_t  balance;
+
+      auto getKey() const { return realBalance(balance, precision); }
+   };
+   PSIO_REFLECT(TokenBalance, symbolId, tokenId, precision, balance);
 
+   struct TokenOwnerDetails
+   {
+      AccountNumber account;
+      AccountNumber authService;
+   };
+   PSIO_REFLECT(TokenOwnerDetails, account, authService);
+
+   // Removed inflation details because the inflation params aren't yet supported
+   //   in the service.
+   struct TokenDetail
+   {
+      TID                id;
+      NID                ownerNft;
+      TokenOwnerDetails  ownerDetails;
+      psibase::Bitset<8> config;
+      Precision          precision;
+      SID                symbolId;
+      Quantity           currentSupply;
+      Quantity           maxSupply;
+   };
+   PSIO_REFLECT(TokenDetail,
+                id,
+                ownerNft,
+                ownerDetails,
+                config,
+                precision,
+                symbolId,
+                currentSupply,
+                maxSupply);
+
+}  // namespace TokenQueryTypes
+
+using namespace TokenQueryTypes;
 struct TokenQuery
 {
    auto allBalances() const
    {  //
-      return tokenService.index<BalanceTable, 0>();
+      return tokenService.open<BalanceTable>().getIndex<0>();
    }
-   auto userBalances(AccountNumber user) const
-   {  //
-      vector<UserBalanceRecord> balances;
-      for (auto tokenType : tokenService.index<TokenTable, 0>())
+
+   /// Provides a way to query all balances for a given user.
+   /// Further filtering can be done to restrict the results by balance.
```

Comment:
I don't understand what it means to compare the balances of different tokens or how this comparison would be useful.

Reply:
A similar function exists on multiple different eos block explorers, for example, because old accounts accumulate small numbers of junk tokens. But tbh it's really not that useful as a query on the server side. I think i'll just delete it. and delete the gt,lt,ge,le flags

Reply:
Deleted